File: informatics/queue.cpp
#include <bits/stdc++.h>

class Node{
  public:
    int data;
    Node* next;
    Node();
};
Node::Node(){
  next = NULL;
}

class Queue{
  public:
    Node* front;
    Node* back;
    void push(int data);
    ~Queue();
    Queue();
    int pop();
    int length;
    int frontValue();
    int backValue();
    int size();
    void clear();
    Node* createNode(int data);
};

Queue::Queue(){
  front = NULL;
  back = NULL;
  length = 0;
}

Queue::~Queue(){
  clear();
}

Node* Queue::createNode(int data){
  Node* node = new Node();
  node->data = data;
  node->next = NULL;
  return node;
}

void Queue::push(int data){
  length++;
  if(back==NULL){
    front = back = createNode(data);
    return;
  }
  Node* node = createNode(data);
  this->back->next = node;
  this->back = node;
}

int Queue::pop(){
  length--;
  if(front==NULL)return INT_MIN;
  int value = front->data;
  Node* tmp = front;
  front = front->next;
  if(front==NULL) back=NULL;
  delete tmp;
  return value;
}

int Queue::frontValue(){
  if(front==NULL)return INT_MIN;
  return front->data;
}

int Queue::backValue(){
  if(back==NULL)return INT_MIN;
  return back->data;
}

int Queue::size(){
  return this->length;
}

Node* deleteNext(Node* node){
  if(node->next != NULL){
    node->next = deleteNext(node->next);
  }
  delete node;
  return NULL;
}

void Queue::clear(){
  if (this->front == NULL){return; }
  this->front = deleteNext(this->front);
  this->back = NULL;
  this->length=0;
}



int main(int argc, char *argv[])
{
  Queue* queue = new Queue();
  std::string str;
  while(std::cin>>str){
    if(str == "push"){
      int data;
      std::cin >> data;
      queue->push(data);
      std::cout << "ok" << std::endl;
    }
    else if(str == "pop"){
      std::cout << queue->pop() << std::endl;
    }
    else if (str == "front"){
      std::cout << queue->frontValue() << std::endl;
    }
    else if (str == "back"){
      std::cout << queue->backValue() << std::endl;
    }
    else if (str == "size"){
      std::cout << queue->size() << std::endl;
    }
    else if (str == "clear"){
      queue->clear();
      std::cout << "ok" << std::endl;
    }
    else if (str == "exit"){
      std::cout << "bye" << std::endl;
      break;
    }
  }
  delete queue;
  return 0;
}

File: dsa/Queue.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;

/**
 * Queue
 */
public class Queue {
  /**
   * Node
   */
  class Node {
    Node next;
    int data;
    public Node(int data) {
      this.data = data;
    }
  }

  public Node head;
  public Node tail;
  public int count;

  public Queue() {
    this.count = 0;
  }

  public void push(int value) {
    count++;
    Node tmp = new Node(value);
    if (this.head == null) {
      this.head = this.tail = tmp;
      return;
    } 
    tail.next = tmp;
    tail = tmp;
  }
  
  public int pop() {
    if (this.head == this.tail) {
      int data = this.head.data;
      this.clear();
      return data;
    }
    int data = this.head.data;
    head = head.next;
    count--;
    return data;
  }

  public int front() {
    return this.head.data; 
  }

  public void clear() {
    this.head = null;
    this.tail = null;
    this.count = 0;
  }

  public static void main(String[] args) throws IOException {
    Queue queue = new Queue();
    StreamTokenizer tf = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));

    while (true) {
      tf.nextToken();
      if (tf.sval.equals("push")) {
        tf.nextToken();
        int value = (int)tf.nval;
        queue.push(value);
        System.out.println("ok");
      }
      else if (tf.sval.equals("pop")) {
        System.out.println(queue.pop());
      }
      else if (tf.sval.equals("front")) {
        System.out.println(queue.front());
      }
      else if (tf.sval.equals("size")) {
        System.out.println(queue.count);
      }
      else if (tf.sval.equals("clear")) {
        queue.clear();
        System.out.println("ok");
      }
      else if (tf.sval.equals("exit")) {
        System.out.println("bye");
        break;
      }
    }
  }
}
